vlm: optimize tensor transport

mickqian opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

While extending, a `TokenizedGenerateReqInput` is expected to be sent between processes (normally `TokenizerManager` -> `Scheduler` ).

For mllms, when `TokenizedGenerateReqInput` carries tensors(e.g., pixel_values), tensor pickling with copy can be expensive. 

Luckily, when the sender and receiver are in the same node, `cuda_ipc` could save these operations.

This pr can reduce TTFT for vlms with multimodal data in request.



<!-- Explain the purpose of this PR and the goals it aims to achieve. -->
## Modifications

1. `TransportableTensor` class with serialize/deserialize functions
2. modify the `__getstate__` and `__setstate__` (which are basically hooks for pickles), to modify the tensor in-place

<!-- Describe the changes made in this PR. -->

## Benchmark

<img width="1004" height="722" alt="transport_proxy_tensor_performance" src="https://github.com/user-attachments/assets/16831f62-a3a2-4d19-8dd4-272b4eec5260" />

mmmu, all accuracy remains the same

| time                             | base                          | pr                                                          |
|----------------------------------|-------------------------------|-------------------------------------------------------------|
| OpenGVLab/InternVL2_5-8B (tp==1) | 207.7 |  173.3 |
| OpenGVLab/InternVL2_5-8B (tp==4) | 158.1       |     139.2                          |
## TODO
1. support nccl transport for those tensors
2. simplify the logic


## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

mickqian committed:
tensor_transport_mode

mickqian committed:
tp=0 and tp>0 working

mickqian commented:
### Req

tp==1

|                        | accuracy(before) | time(before) | accuracy(after) | time(after) |
|------------------------|------------------|--------------|-----------------|-------------|
| Gemma-3-4b-it          | 0.384            | 226.2        | 0.384           | 219.9       |
| Qwen2.5-VL-7B-Instruct | 0.467            | 352.7        | 0.467           | 349.8       |
| Minicpmv               | 0.436            | 232.5        | 0.436           | 224.1       |

tp==4

|                        | accuracy(before) | time(before) | accuracy(after) | time(after) |
|------------------------|------------------|--------------|-----------------|-------------|
| Gemma-3-4b-it          | 0.386            | 265.2        | 0.386           | 264.7       |
| Qwen2.5-VL-7B-Instruct | 0.467            | 391.9        | 0.472           | 367.9       |
| Minicpmv               | 0.436            | 225.1        | 0.438           | 227.2       |

mickqian committed:
working

mickqian committed:
fallback on tp

mickqian committed:
cleanup

mickqian committed:
cleanup

mickqian committed:
refactor with TransportableTensor

mickqian committed:
work with tp

mickqian committed:
update

mickqian committed:
remove deepcopy

mickqian committed:
cleanup

mickqian committed:
cleanup

mickqian committed:
refactor: unify feature field of MultimodalDataItem

mickqian committed:
update

mickqian committed:
update

mickqian committed:
update

mickqian committed:
update

mickqian committed:
update

mickqian committed:
replace awkward __getstate__

mickqian committed:
update

mickqian committed:
cleanup

mickqian committed:
cleanup

mickqian committed:
revert mmmu related

mickqian committed:
revert mmmu related

mickqian committed:
cleanup

mickqian committed:
processor

mickqian committed:
TransportableTensor __getstate__

yhyang201 commented:
/gemini review

mickqian committed:
rename

mickqian committed:
update

mickqian committed:
fix

JustinTong0323 commented on python/sglang/srt/multimodal/processors/base_processor.py:
```diff
@@ -548,6 +546,11 @@ def collect_mm_items_from_processor_output(
                     attr_name = "feature"
+                    # replace the feature tensor with a proxy
+                    if isinstance(value, torch.Tensor):
+                        value = TransportProxyTensor(
```
Maybe better to have this manipulation occur within the mmdataitem to enhance extensibility in the future?

JustinTong0323 commented on python/sglang/utils.py:
```diff
@@ -28,6 +29,24 @@
+
+
+@execute_once
+def info_once(message: str):
```
where we use it?

JustinTong0323 reviewed (approved):
Overall LGTM

mickqian committed:
upd

mickqian committed:
upd

JustinTong0323 committed:
Merge branch 'main' into optimize_req

JustinTong0323 committed:
remove duplicate multimodal_processors/qwen_audio.py

JustinTong0323 committed:
refactor: update terminology from 'precomputed features' to 'precomputed embeddings' in relevant classes and tests

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
refactor: update constructors in multimodal processors to accept additional arguments

Updated the constructors of various multimodal processor classes to include *args and **kwargs, allowing for more flexible initialization and compatibility with future extensions.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 commented on python/sglang/srt/multimodal/processors/base_processor.py:
```diff
@@ -624,4 +623,19 @@ def process_and_combine_mm_data(
+                )
+            elif (
+                isinstance(item.precomputed_embeddings, torch.Tensor)
+                and item.feature.is_cuda
```
Does `item.precomputed_embeddings` seem to be mutually exclusive with `item.feature`?

mickqian committed:
fix

JustinTong0323 committed:
fix: rename precomputed_features to precomputed_embeddings

JustinTong0323 committed:
Merge branch 'main' into optimize_req
